[WIP] Adding a winsorizer transformer

rtalrejanyc opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

Fixes #3130.

#### What does this implement/fix? Explain your changes.

This implements a winsorizer transformer for clipping feature values.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

rtalrejanyc committed:
Adding a winsorizer transformer.

rtalrejanyc committed:
Adding a winsorizer transformer.

rtalrejanyc committed:
Adding a winsorizer transformer.

rtalrejanyc committed:
Adding a winsorizer transformer.

amueller commented:
Thanks. As mentioned it would be cool if you could add an example on when this is useful.

rtalrejanyc committed:
Added a doctest.

rtalrejanyc committed:
Added a doctest.

rtalrejanyc committed:
Added a doctest.

rtalrejanyc committed:
Added a doctest.

rtalrejanyc committed:
Adding an example for Winsorizer.

rtalrejanyc committed:
Adding an example for Winsorizer.

rtalrejanyc committed:
Adding an example for Winsorizer.

rtalrejanyc committed:
Added more explanation to winsorizer example. Also, moved class
Winsorizer to a more logical location in the file.
Removed tabs and changed a docstring.

rtalrejanyc committed:
Added more explanation to winsorizer example. Also, moved class
Winsorizer to a more logical location in the file.

rtalrejanyc committed:
Added more explanation to winsorizer example. Also, moved class
Winsorizer to a more logical location in the file.

rtalrejanyc committed:
Added more explanation to winsorizer example. Also, moved class
Winsorizer to a more logical location in the file.

rtalrejanyc committed:
Adding test for Winsorizer.

rtalrejanyc committed:
Adding test for Winsorizer.

rtalrejanyc committed:
Adding test for Winsorizer.

Removed tabs and changed a docstring.

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2901,6 +2902,95 @@ def power_transform(X, method='box-cox', standardize=True, copy=True):
 
 
+class Winsorizer(BaseEstimator, TransformerMixin):
+    """Transforms each feature by clipping from below at the pth quantile
```
"Clips each feature at the pth quantile from above and below"

should be sufficient.

Add Read more in the User Guide (as in most other estimator docstrings) once a section is in the user guide.

jnothman commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2465,3 +2466,23 @@ def test_power_transformer_lambda_zero():
+    ub = np.percentile(X, 100 * (1 - quantile), axis=0)
+    X_expected = np.clip(X, lb, ub)
+
+    assert_array_almost_equal(X_trans, X_expected)
```
Please add a test for copy.

jnothman commented:
Add some motivation (and reference the example) in doc/modules/preprocessing.rst.

Not sure if this should be added to plot_all_scaling...

amueller commented:
I suggested adding something to the linear model outlier examples. I think ``plot_all_scaling`` is a good idea, too, though.

rtalrejanyc commented:
Thanks for the comments. I think an example would fit more naturally in
plot_all_scaling. I hadn't seen this page before. I'll make the changes.

On Sat, Jun 16, 2018 at 1:57 PM, Andreas Mueller <notifications@github.com>
wrote:

> I suggested adding something to the linear model outlier examples. I think
> plot_all_scaling is a good idea, too, though.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/11278#issuecomment-397813783>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AWjMQwLRnZMNkJzxnBhJHCd_phgk8ma1ks5t9Q7bgaJpZM4Up3a8>
> .
>

rtalrejanyc committed:
Merge remote-tracking branch 'upstream/master' into winsorizer

rtalrejanyc committed:
Removed tabs and changed a docstring.

rtalrejanyc committed:
Removed tabs and changed a docstring.

rtalrejanyc committed:
Merge branch 'winsorizer' of https://github.com/rtalrejanyc/scikit-learn into winsorizer

sklearn-lgtm commented:
This pull request **introduces 8 alerts** when merging a4e6ceb37d402e182295a551aa86bea45d4dbcab into 1ff836438775cca5c480bde82b323651046c3fb2 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-446efa08af55a2917da899d1f5bb99828d4a5d11)

**new alerts:**

* 3 for Comparison of identical values
* 2 for Variable defined multiple times
* 1 for Unused local variable
* 1 for Unused import
* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

jnothman commented:
That merge wasn't right. Please consider using `git reset --hard ref` where ref is the last good commit.

rtalrejanyc committed:
Fixing bad merge.

sklearn-lgtm commented:
This pull request **introduces 6 alerts** when merging da03db41323089564a0d262b4f6aafaa4d695c0a into 1ff836438775cca5c480bde82b323651046c3fb2 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-9fba8752296137f2f7244a743f30178c6b0135a6)

**new alerts:**

* 3 for Comparison of identical values
* 1 for Unused local variable
* 1 for Unused import
* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

jnothman commented:
Still broken

rtalrejanyc committed:
Merge branch 'winsorizer' of https://github.com/rtalrejanyc/scikit-learn into winsorizer

sklearn-lgtm commented:
This pull request **introduces 6 alerts** when merging fb3cb7c1256a51b641cb8bc09b26fa090815181c into 1ff836438775cca5c480bde82b323651046c3fb2 - [view on LGTM.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-e2b5e8de961d3bf9a82c688c3732868c7fa5107e)

**new alerts:**

* 3 for Comparison of identical values
* 1 for Unused local variable
* 1 for Unused import
* 1 for Module is imported with &#39;import&#39; and &#39;import from&#39;

---

*Comment posted by [LGTM.com](https://lgtm.com)*

jnothman commented:
please stop modifying this branch and I'll try help you clean it up.​

jnothman commented:
Actually, I think this is quite a mess, and you might be better off just copying the relevant portions into a new branch and starting afresh.

rtalrejanyc commented:
Yes, looks like I’ll need to start from scratch. Thanks!

Sent from my iPhone

> On Jun 22, 2018, at 12:08 AM, Joel Nothman <notifications@github.com> wrote:
> 
> Actually, I think this is quite a mess, and you might be better off just copying the relevant portions into a new branch and starting afresh.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

theadammurphy commented:
Was this ever solved or implemented?

haiatn commented:
> Was this ever solved or implemented?

I would have loved it if someone would have implemented it but I think no one had the time

adrinjalali commented:
Closing since it'd need a new start anyway.
